### PR TITLE
[MAINTENANCE] move tool config to pyproject.toml - fork

### PR DESCRIPTION
**Pull request by Kilo59**
Changes proposed in this pull request:

Splitting from #5506 
* Creating a [pyproject.toml file](https://peps.python.org/pep-0518/) to define the build system ge uses and start centralizing tool config to it (black, isort, pytest, etc.)]
* Removing run time args from `black` and `isort`. Pulling config from `pyproject.toml` will force consistent usage of the tools in `ci`, local dev environments and via `pre-commit` hooks. 

Here's some lists of tools that support a `pyroject.toml` config
https://github.com/carlosperate/awesome-pyproject
https://github.com/flying-sheep/awesome-python-packaging

### TODO
- [x] resolve potential `versioneer` issue
- [x] remove runtime args for `isort` and `black`
  - [x] set `black` & `isort` config in `pyproject.toml`, prevent the need to keep various `ci` & `pre-commit` runtime parameters in sync. 
- [x] update `isort` version

### Notable References
https://setuptools.pypa.io/en/latest/build_meta.html

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- ~I have commented my code, particularly in hard-to-understand areas~
- ~I have made corresponding changes to the documentation~
- ~I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.~
- [x] ~I have run any local integration tests~ Reinstalled `great_expectations` locally and made sure that nothing is broken.

### CODE REVIEW

**Comment by Kilo59** [comment]
There was some good questions and discussion in #5506 I will be linking to the relevant comments below.

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,30 @@
+[build-sytem]
+requires = ["setuptools", "wheel"]
+build-backend = "setuptools.build_meta"
```
Comment (Kilo59) [inline pyproject.toml]: @billdirks 
Question about the build-system settings.

https://github.com/great-expectations/great_expectations/pull/5506#discussion_r921784934

>How/where is this file read in?

This gets read by pip.
https://peps.python.org/pep-0518/#build-system-table

File: .isort.cfg [hunk 2]
```diff
@@ -1,4 +0,0 @@
-[settings]
-profile=black
-known_first_party=great_expectations,importlib_metadata # Added here because seed-isort-config does not recognize it
-known_third_party=IPython,_pytest,altair,black,boto3,botocore,click,dateutil,freezegun,ipywidgets,jinja2,jsonpatch,jsonschema,marshmallow,mistune,mock,moto,nbconvert,nbformat,numpy,pandas,pkg_resources,pyparsing,pytest,requests,requirements,ruamel,scipy,setuptools,sphinx,sqlalchemy,tzlocal
```
Comment (Kilo59) [inline .isort.cfg:4]: I have left these settings out of the `pyproject.toml` config.
Changes as of `isort 5` make these no longer necessary. 

Screenshot below showing isort being run with these setting outside of the virtual environment without resulting in any sorting changes/regressions.

image

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+junit_family="xunit2"
+markers = [
+    "docs: mark a test as a docs test.",
+    "unit: mark a test as a unit test.",
+    "integration: mark test as an integration test.",
+    "external_sqldialect: mark test as requiring install of an external sql dialect.",
+    "cloud: mark test as being relevant to Great Expectations Cloud.",
+    "base_data_context: mark test as being relevant to BaseDataContext, which will be removed during refactor",
+]
+testpaths = "tests"
```
Comment (Kilo59) [inline pyproject.toml]: One thing we might one to add is some guidance to contributors/developers suggesting that they add any new tool configuration to this file if its supported.
Not sure if we need that though or where to put it.

File: .gitignore [hunk 4]
```diff
@@ -137,3 +137,7 @@ DO_NOT_COMMIT
 
 # Hive Metastore from Spark
 metastore_db
+
+# dependency managment
+Pipfile
+Pipfile.lock
```
Comment (Shinnnyshinshin) [inline .gitignore]: Just learned more about this [here](https://stackoverflow.com/questions/46330327/how-are-pipfile-and-pipfile-lock-used). Love the change

**Review by Shinnnyshinshin — APPROVED** [review]
Thank you for introducing this change to @Kilo59. Love how the `toml` file is laid out, and looking forward to using it more ✅

**Comment by Kilo59** [comment]
Couple points I just want to bring up related to this change.

A previous minor version of `pip` did lead to broken installs when combined with `pyproject.toml` files and certain `setup.py` usage. Specifically when folks imported their own package from within `setup.py`, however we don't do this, so it shouldn't be a problem.
https://github.com/pypa/pip/issues/6163

One thing we may need to test and make adjustments for is our usage of `versioneer`.
See the example below for how `pandas` deals with it. 
https://github.com/pandas-dev/pandas/blob/main/pyproject.toml#L10-L12
https://github.com/python-versioneer/python-versioneer/issues/193

Edit:
~Marking PR as `draft` until I have had a chance to thoroughly test the issue with `versioneer`.~
I wasn't able to reproduce the `versioneer` problem locally, but decided to copy what `pandas` does just to be safe.

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -13,15 +13,13 @@ repos:
       - id: requirements-txt-fixer
         exclude: docs/.*
   - repo: https://github.com/timothycrosley/isort
-    rev: 5.5.2
+    rev: 5.10.1
     hooks:
       - id: isort
-        exclude: venv/.*\|docs/.*|tests/test_sets/broken_excel_file\.xls
```
Comment (Kilo59) [inline .pre-commit-config.yaml:19]: Any idea why we were telling the `isort` hook to ignore this excel file?

**Comment by Kilo59** [comment]
Getting lots of environmental errors that might be caused by this PR originating from a fork of `gx`.

Opening a fresh PR that doesn't originate from my fork.
If this works I will close this PR in favor of https://github.com/great-expectations/great_expectations/pull/5649

**Pull request closed without merging**